src/query/storage/promremote/query_coverter.go

```diff
@@ -0,0 +1,77 @@
+// Copyright (c) 2021  Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package promremote
+
+import (
+	"time"
+
+	"github.com/golang/snappy"
+	"github.com/pkg/errors"
+	"github.com/prometheus/prometheus/prompb"
+
+	"github.com/m3db/m3/src/query/storage"
+)
+
+var errNilQuery = errors.New("received nil query")
+
+func convertAndEncodeWriteQuery(query *storage.WriteQuery) ([]byte, error) {
+	if query == nil {
+		return []byte{}, errNilQuery
+	}
+	promQuery := convertWriteQuery(query)
+	data, err := promQuery.Marshal()
+	if err != nil {
+		return nil, err
+	}
+	return snappy.Encode(nil, data), nil
+}
+
+func convertWriteQuery(query *storage.WriteQuery) *prompb.WriteRequest {
+	if query == nil {
+		return nil
+	}
+
+	ourLabels := storage.TagsToPromLabels(query.Tags())
+	labels := make([]prompb.Label, len(ourLabels))
+	for i, tag := range ourLabels {
+		labels[i] = prompb.Label{
+			Name:  string(tag.Name),
+			Value: string(tag.Value),
+		}
+	}
+
+	samples := make([]prompb.Sample, len(query.Datapoints()))
+	for i, dp := range query.Datapoints() {
+		samples[i] = prompb.Sample{
+			Value:     dp.Value,
+			Timestamp: dp.Timestamp.ToNormalizedTime(time.Millisecond),
+		}
+	}
```

Comment:
nit: Again prefer and more consistent in M3 code base to do `foo := make(bar, 0, len(expected))` and `foo = append(foo, bar{...})` to avoid code bugs that can creep in more easily with the index based approach when code is later modified with early return and/or continue/etc.